### PR TITLE
Fix health checking for ephemeral poststart tasks

### LINKED ISSUE #9254: Both tasks marked as unhealthy if only one fails.
# Nomad version

0.12.5

### Operating system and Environment details

MacOS Catalina 10.15.4, but I'm experiencing similar issues running through OpenStack.

### Issue

When a task group has 2+ tasks and _one_ fails, both tasks are marked unhealthy with `Task not running for min_healthy_time of 10s by deadline` in the UI (see reproduction steps below). Maybe this is intentional for some good reason, but the UX is confusing regardless.

nomad-both-tasks-fail

### Reproduction steps

Please see source and instructions in [erichulburd/consul-nomad-issues](https://github.com/erichulburd/consul-nomad-issues).

### Job file (if appropriate)

[erichulburd/consul-nomad-issues/job.json](https://github.com/erichulburd/consul-nomad-issues/blob/master/job.json).

### LINKED ISSUE #10058: unhealthy deployment with poststart lifecycle
### Nomad version
Nomad v1.0.3 (08741d9f2003ec26e44c72a2c0e27cdf0eadb6ee)

### Operating system and Environment details
Ubuntu 18.04.5 LTS

### Issue
Deployement with poststart task is immediately marked as unhealthy and still running as an active deployment until deadline hits.
The Service is registered healthy in Consul and reachable via traefik. 

nomad_poststart_1

nomad_poststart_2

Removing the poststart job and the deployment is running fine.

I guess the poststart job is exiting too fast (under 10s) und therefore hits the min_healthy_time of 10s(default).

### Job file

```
job "nginx" {

  datacenters = ["prod"]
  namespace   = "default"
  type        = "service"

  group "nginx" {
    count = 1
    service {
      name = "${BASE}"
      tags = [
        "traefik.enable=true",
        "traefik.http.routers.nginx.entrypoints=websecure",
        "traefik.http.routers.nginx.rule=Host(`nginx.domain`)",
        "traefik.http.routers.nginx.tls=true"
      ]
      port = "http"
      check {
        name     = "alive"
        type     = "tcp"
        interval = "2s"
        timeout  = "2s"
      }
    }

    network {
      port "http" {
        to = 80
      }
    }

    update {
      min_healthy_time  = "10s"
      healthy_deadline  = "5m"
      progress_deadline = "10m"
    }

    task "nginx-web" {
      driver = "docker"
      config {
        image = "nginx:latest"
        ports = ["http"]
      }
    }

    task "nginx-poststart" {
      driver = "docker"
      config {
        image   = "alpine:latest"
        command = "sh"
        args    = ["-c", "sleep 5 && echo 'running....' && exit 0;"]

      }
      lifecycle {
        hook    = "poststart"
        sidecar = false
      }
    }
  }
}
```

Enforcing the poststart job to wait for the min_healthy_time from the web service and exiting after, fixed the issue. 
So i set the sleep timer slightly above the **min_healthy_time = "10s"** and it works.

```
    task "nginx-poststart" {
      driver = "docker"
      config {
        image   = "alpine:latest"
        command = "sh"
        args    = ["-c", "sleep 12 && echo 'running....' && exit 0;"]

      }
      lifecycle {
        hook    = "poststart"
        sidecar = false
      }
    }
```
nomad_poststart_3

Seen in [9361](https://github.com/hashicorp/nomad/issues/9361)

Is it a bug or working as expected ? 

Thanks.

### PR DESCRIPTION
**Pull request by beautifulentropy**
This PR updates the logic in the Nomad client's alloc health tracker which
erroneously marks existing healthy allocations with dead poststart ephemeral
tasks as unhealthy even if they were already successful during a previous
deployment. For repro see: https://github.com/hashicorp/nomad/issues/10058#issuecomment-1020658228

Currently, users are having to insert sleep after short-lived ephemeral tasks.
This change ensures that poststart lifecycle tasks which have succeeded,
even before 'min_healthy_time', will not result in the whole allocation being
marked as unhealthy.

I've attempted to include test coverage that sticks to your existing conventions.

- Tracker will not attempt to evaluate the health of poststart ephemeral tasks
  which have already succeeded with any running duration.
- Add a mock and helper for deployments of allocations with poststart tasks

Fixes #9254
Fixes #10058

### CODE REVIEW

File: client/allochealth/tracker.go [hunk 1]
```diff
@@ -277,15 +277,35 @@ func (t *Tracker) watchTaskEvents() {
...
+					// 'ignore_min_healthy_time' to 'true', skip evaluation.
+					if t.lifecycleTasks[taskName].IgnoreMinHealthyTime {
+						continue
+					}
+				}
 			}
 
 			// One of the tasks has failed so we can exit watching
-			if state.Failed || (!state.FinishedAt.IsZero() && t.lifecycleTasks[taskName] != structs.TaskLifecycleHookPrestart) {
+			if state.Failed || (!state.FinishedAt.IsZero() && t.lifecycleTasks[taskName].Hook != structs.TaskLifecycleHookPrestart) {
```
Comment (tgross) [inline client/allochealth/tracker.go:308]: Hi @beautifulentropy! We're going to take some time to review this in more depth but right now a bunch of tests are failing with a nil pointer dereference here:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1399289]

goroutine 134 [running]:
github.com/hashicorp/nomad/client/allochealth.(*Tracker).watchTaskEvents(0xc0007c1a40)
	/home/circleci/go/src/github.com/hashicorp/nomad/client/allochealth/tracker.go:308 +0xe49
created by github.com/hashicorp/nomad/client/allochealth.(*Tracker).Start
	/home/circleci/go/src/github.com/hashicorp/nomad/client/allochealth/tracker.go:139 +0x5f
```

I think that's because the `t.lifecycleTasks` might not have a key for `taskName`, which was safe before this change because the value of `t.lifecycleTasks["nonexistenttask"]` would just be an empty string we could compare. But now it's a pointer and dereferencing it to get the `.Hook` field panics if it's nil.

Comment (beautifulentropy) [reply client/allochealth/tracker.go]: Ah, thanks! I've gone ahead and addressed this.

File: client/allochealth/tracker.go [hunk 2]
```diff
@@ -277,17 +277,37 @@ func (t *Tracker) watchTaskEvents() {
...
+				// If this is a poststart task which has already succeeded we
+				// want to check for two possible success conditions before
+				// attempting to evaluate it.
+				if t.lifecycleTasks[taskName].Hook == structs.TaskLifecycleHookPoststart && state.Successful() {
+
+					// If the task was successful and it's runtime is at least
+					// t.minHealthyTime, skip evaluation.
+					if state.FinishedAt.Sub(state.StartedAt) >= t.minHealthyTime {
+						continue
+					}
```
Comment (tgross) [inline client/allochealth/tracker.go:297]: This tracker has an existing special case for _pre_ start tasks, and I wonder if we should be using that here for poststart as well.

All tasks with lifecycle hooks without the `sidecar = true` field are supposed to be tolerated already as "short-lived ephemeral tasks" without any additional configuration:
* A sidecar prestart task needs to start up and keep running. If it exits for any reason, that's a failure.
* A non-sidecar prestart task needs to finish. It's only a failure if it exits with a non-0 exit code.

So rather than introducing a new `ignore_min_healthy_time` field, couldn't we use the same conditional here for poststart that we're using for prestart tasks? (I admit I could be missing something here for sure.)

Comment (beautifulentropy) [reply client/allochealth/tracker.go:297]: Ah, I think you're absolutely correct here. I've made the appropriate changes inside of the lifecycleTask loop and reverted my addition of `ignore_min_healthy_time`.

I did a little bit of testing and this has corrected both the 'not running for min_healthy_time' bug and the erroneous marking of previous allocations as unhealthy.

Comment (beautifulentropy) [reply client/allochealth/tracker.go:297]: > A non-sidecar prestart task needs to finish. It's only a failure if it exits with a non-0 exit code.

I updated the docstring for the `Successful` method of `structs.TaskState` to reflect this.

**Review by tgross — APPROVED** [review]
LGTM! Thanks @beautifulentropy!

**Comment by tgross** [comment]
This will ship in the next major release (1.3.0) and get backported to 1.2.x and 1.1.x.
